File: backend/api/discovery.py
"""Discovery API — pywebview bridge for remote server discovery and connection."""

from __future__ import annotations

import logging
from typing import Any

from backend.services.discovery_service import DiscoveryService
from backend.services.inference_manager import InferenceServiceManager

logger = logging.getLogger(__name__)


class DiscoveryAPI:
    """pywebview js_api mixin for remote server discovery."""

    def __init__(self, manager: InferenceServiceManager, discovery: DiscoveryService) -> None:
        self._manager = manager
        self._discovery = discovery

    def get_discovered_servers(self) -> dict[str, Any]:
        """Return the list of inference servers found on the LAN."""
        try:
            servers = self._discovery.get_servers()
            return {"status": "success", "servers": servers}
        except Exception as e:
            logger.error("Failed to get discovered servers: %s", e)
            return {"status": "error", "message": str(e)}

    def connect_to_server(self, host: str, port: int, **kwargs) -> dict[str, Any]:
        """Connect to a remote inference server."""
        try:
            self._manager.connect_remote(host, port)
            return {"status": "success", "mode": "remote"}
        except Exception as e:
            logger.error("Failed to connect to %s:%d: %s", host, port, e)
            return {"status": "error", "message": str(e)}

    def disconnect_from_server(self) -> dict[str, Any]:
        """Disconnect from the remote server and revert to local mode."""
        try:
            self._manager.disconnect_remote()
            return {"status": "success", "mode": "local"}
        except Exception as e:
            logger.error("Failed to disconnect: %s", e)
            return {"status": "error", "message": str(e)}

    def get_connection_mode(self) -> dict[str, Any]:
        """Return the current connection mode and connected server info."""
        try:
            return {
                "status": "success",
                "mode": self._manager.mode,
                "connected_server": self._manager.connected_server,
            }
        except Exception as e:
            logger.error("Failed to get connection mode: %s", e)
            return {"status": "error", "message": str(e)}

File: frontend/src/stores/discovery.ts
import { defineStore } from 'pinia';
import { getApiAsync } from '@/bridge';
import type { DiscoveredServer } from '@/types/inference';

let _pollTimer: ReturnType<typeof setInterval> | null = null;

export const useDiscoveryStore = defineStore('discovery', {
  state: () => ({
    servers: [] as DiscoveredServer[],
    connectionMode: 'local' as string,
    connectedServer: null as { host: string; port: number } | null,
    connecting: false,
    error: null as string | null,
  }),

  getters: {
    isRemote(state): boolean {
      return state.connectionMode === 'remote';
    },
  },

  actions: {
    async refreshServers() {
      try {
        const api = await getApiAsync();
        const res = await api.get_discovered_servers();
        if (res.status === 'success' && res.servers) {
          this.servers = res.servers;
        }
      } catch {
        // Discovery is best-effort
      }
    },

    async loadConnectionMode() {
      try {
        const api = await getApiAsync();
        const res = await api.get_connection_mode();
        if (res.status === 'success') {
          this.connectionMode = res.mode ?? 'local';
          this.connectedServer = res.connected_server ?? null;
        }
      } catch {
        // Non-critical
      }
    },

    async connectToServer(host: string, port: number) {
      if (this.connecting) return;
      this.connecting = true;
      this.error = null;
      try {
        const api = await getApiAsync();
        const res = await api.connect_to_server({ host, port });
        if (res.status === 'success') {
          this.connectionMode = res.mode ?? 'remote';
          this.connectedServer = { host, port };
        } else {
          this.error = res.message ?? 'Failed to connect';
        }
      } catch (e: any) {
        this.error = e.message ?? 'Failed to connect';
      } finally {
        this.connecting = false;
      }
    },

    async disconnectFromServer() {
      this.error = null;
      try {
        const api = await getApiAsync();
        const res = await api.disconnect_from_server();
        if (res.status === 'success') {
          this.connectionMode = res.mode ?? 'local';
          this.connectedServer = null;
        } else {
          this.error = res.message ?? 'Failed to disconnect';
        }
      } catch (e: any) {
        this.error = e.message ?? 'Failed to disconnect';
      }
    },

    startDiscoveryPolling() {
      if (_pollTimer) return;
      this.refreshServers();
      _pollTimer = setInterval(() => this.refreshServers(), 3000);
    },

    stopDiscoveryPolling() {
      if (_pollTimer) {
        clearInterval(_pollTimer);
        _pollTimer = null;
      }
    },
  },
});
